ext/nio4r/monitor.c

```diff
@@ -117,6 +119,34 @@ static VALUE NIO_Monitor_initialize(VALUE self, VALUE io, VALUE interests, VALUE
     return Qnil;
 }
 
+static VALUE NIO_Monitor_setInterests(VALUE self, VALUE interests){
+    //Check Whether the Monitor is already Closed?
+    if(NIO_Monitor_is_closed(self) == Qfalse){
+        struct NIO_Monitor *monitor;
+        ID interests_id;
+
+        interests_id = SYM2ID(interests);
+        Data_Get_Struct(self, struct NIO_Monitor, monitor);
+
+        if(interests_id == rb_intern("r")) {
+            monitor->interests = EV_READ;
+        } else if(interests_id == rb_intern("w")) {
+            monitor->interests = EV_WRITE;
+        } else if(interests_id == rb_intern("rw")) {
+            monitor->interests = EV_READ | EV_WRITE;
+        } else {
+            rb_raise(rb_eArgError, "invalid event type %s (must be :r, :w, or :rw)",
+                RSTRING_PTR(rb_funcall(interests, rb_intern("inspect"), 0, 0)));
+        }
+
+        rb_ivar_set(self, rb_intern("interests"), interests);
```

Comment:
k. I'll look into this and correct it.

Upekshe Jayasekera
Undergraduate
Department of Computer Science and Engineering
University of Moratuwa
Sri Lanka

On 8 June 2015 at 00:06, Tony Arcieri <notifications@github.com> wrote:

> In ext/nio4r/monitor.c
> <https://github.com/celluloid/nio4r/pull/73#discussion_r31879290>:
>
> > +
> > +        interests_id = SYM2ID(interests);
> > +        Data_Get_Struct(self, struct NIO_Monitor, monitor);
> > +
> > +        if(interests_id == rb_intern("r")) {
> > +            monitor->interests = EV_READ;
> > +        } else if(interests_id == rb_intern("w")) {
> > +            monitor->interests = EV_WRITE;
> > +        } else if(interests_id == rb_intern("rw")) {
> > +            monitor->interests = EV_READ | EV_WRITE;
> > +        } else {
> > +            rb_raise(rb_eArgError, "invalid event type %s (must be :r, :w, or :rw)",
> > +                RSTRING_PTR(rb_funcall(interests, rb_intern("inspect"), 0, 0)));
> > +        }
> > +
> > +        rb_ivar_set(self, rb_intern("interests"), interests);
>
> You will also need to inform libev that the interests changed, since they
> wind up getting registered directly with the kernel when using backends
> like epoll or kqueue.
>
> They initially get set here:
>
> https://github.com/celluloid/nio4r/blob/master/ext/nio4r/monitor.c#L100
>
> I believe you can use ev_io_set to change them:
>
> http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#ANATOMY_OF_A_WATCHER
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/celluloid/nio4r/pull/73/files#r31879290>.
>